File: WebClient SDK/src/default/ui/cardFiles/customFilesViewCardControlInitializationStrategy.ts
import { LocalizationManager } from 'tessa/localization';
import { SchemeType } from 'tessa/scheme';
import { ViewControlInitializationContext, FileListViewModel } from 'tessa/ui/cards/controls';
import { IViewMetadata, ViewColumnMetadata, ViewMetadataSealed } from 'tessa/views/metadata';
import { CardFilesDataProvider } from './cardFilesDataProvider';
import { CustomCardFilesDataProvider } from './customCardFilesDataProvider';
import { FilesViewCardControlInitializationStrategy } from './filesViewCardControlInitializationStrategy';

/**
 * Custom strategy that provide customized data for files in view.
 */
export class CustomFilesViewCardControlInitializationStrategy extends FilesViewCardControlInitializationStrategy {
  /**
   * Customize view metadata (base + new column)
   * @param context initialization context
   * @returns view metadata
   */
  protected createViewMetadata(context: ViewControlInitializationContext): IViewMetadata {
    const viewMetadata = super.createViewMetadata(context);

    const dateColumnMetadata = new ViewColumnMetadata();
    dateColumnMetadata.caption = LocalizationManager.instance.localize(
      '$CardTypes_Columns_Controls_Date'
    );
    dateColumnMetadata.alias = 'Date';
    dateColumnMetadata.schemeType = SchemeType.DateTime;
    dateColumnMetadata.disableGrouping = true;
    dateColumnMetadata.sortBy = 'Date';

    const descriptionColumnMetadata = new ViewColumnMetadata();
    descriptionColumnMetadata.caption = LocalizationManager.instance.localize(
      '$CardTypes_Controls_Description'
    );
    descriptionColumnMetadata.alias = 'Description';
    descriptionColumnMetadata.schemeType = SchemeType.NullableString;
    descriptionColumnMetadata.disableGrouping = true;
    descriptionColumnMetadata.sortBy = 'Description';

    viewMetadata.columns.set(dateColumnMetadata.alias, dateColumnMetadata);
    viewMetadata.columns.set(descriptionColumnMetadata.alias, descriptionColumnMetadata);
    return viewMetadata;
  }

  /**
   * Customize data provider.
   * @param context initialization context
   * @param viewMetadata prepared view metadata
   * @param fileControl file view model
   * @returns customized card data provider
   */
  protected createDataProvider(
    context: ViewControlInitializationContext,
    viewMetadata: ViewMetadataSealed,
    fileControl: FileListViewModel
  ): CardFilesDataProvider {
    // dummy use for suspend a warning.
    context;
    return new CustomCardFilesDataProvider(viewMetadata, fileControl);
  }
}

File: Source/Extensions/Tessa.Extensions.Default.Client/UI/CardFiles/CustomFilesViewCardControlInitializationStrategy.cs
#nullable enable

using System.Threading.Tasks;
using Tessa.Localization;
using Tessa.Platform.Collections;
using Tessa.Scheme;
using Tessa.UI.Cards.Controls;
using Tessa.UI.Files;
using Tessa.UI.Menu;
using Tessa.Views;
using Tessa.Views.Metadata;

namespace Tessa.Extensions.Default.Client.UI.CardFiles
{
    /// <summary>
    /// Customized file view initialization strategy.
    /// </summary>
    public sealed class CustomFilesViewCardControlInitializationStrategy :
        FilesViewCardControlInitializationStrategy
    {
        #region Constructor

        /// <summary>
        /// Create new instance of custom files in view extension card control initialization strategy. 
        /// </summary>
        /// <param name="viewService"><inheritdoc cref="IViewService" path="/summary"/></param>
        /// <param name="createMenuContextFunc"><inheritdoc cref="CreateMenuContextFunc" path="/summary"/></param>
        /// <param name="contentItemsFactory"><inheritdoc cref="IViewCardControlContentItemsFactory" path="/summary"/></param>
        public CustomFilesViewCardControlInitializationStrategy(
            IViewService viewService,
            CreateMenuContextFunc createMenuContextFunc,
            IViewCardControlContentItemsFactory contentItemsFactory)
            : base(viewService, createMenuContextFunc, contentItemsFactory)
        {
        }

        #endregion

        #region Base Overrides

        /// <inheritdoc/>
        protected override async ValueTask<IViewMetadata> CreateViewMetadataAsync(CardViewControlInitializationContext context)
        {
            var viewMetadata = await base.CreateViewMetadataAsync(context).ConfigureAwait(false);
            var dateColumnMetadata = new ViewColumnMetadata
            {
                Caption = LocalizationManager.Localize("$CardTypes_Columns_Controls_Date"),
                Alias = "Date",
                SchemeType = SchemeType.DateTime,
                DisableGrouping = true,
                SortBy = "Date",
            };
            var descriptionColumnMetadata = new ViewColumnMetadata
            {
                Caption = LocalizationManager.Localize("$CardTypes_Controls_Description"),
                Alias = "Description",
                SchemeType = SchemeType.NullableString,
                DisableGrouping = true,
                SortBy = "Description",
            };
            viewMetadata.Columns.AddRange(dateColumnMetadata, descriptionColumnMetadata);
            return viewMetadata;
        }

        /// <inheritdoc/>
        protected override ValueTask<IDataProvider> CreateDataProviderAsync(
            CardViewControlInitializationContext context, IViewMetadata viewMetadata, IFileControl fileControl) =>
            new(new CustomCardFilesDataProvider(viewMetadata, fileControl));

        #endregion
    }
}
